### ISSUE TITLE
mysql is in blocked state after charm container restart

### ISSUE THREAD

**Issue by hemanthnakkina** [canonical/mysql-k8s-operator#107]
### Bug Description

mysql unit is in blocked state with message "Unable to configure instance"

```
$ juju status mysql
Model      Controller          Cloud/Region        Version  SLA          Timestamp
openstack  microk8s-localhost  microk8s/localhost  3.0-rc1  unsupported  12:42:27+05:30

App    Version  Status   Scale  Charm      Channel  Rev  Address         Exposed  Message
mysql           waiting      1  mysql-k8s  edge      18  10.152.183.210  no       installing agent

Unit      Workload  Agent  Address       Ports  Message
mysql/0*  blocked   idle   10.1.107.255         Unable to configure instance
```

After some debugging of logs, mysql-k8s charm started initializing db/configuring instance and the container charm got restarted during initialisation process (because of 30sec healthcheck). The charm container again gets pebble_ready event after restart and tries to re-initialize DB which fails.

```
2022-10-18T07:05:29.269607Z 0 [ERROR] [MY-010457] [Server] --initialize specified but the data directory has files in it. Aborting.
2022-10-18T07:05:29.269615Z 0 [ERROR] [MY-013236] [Server] The designated data directory /var/lib/mysql/ is unusable. You can remove all files that the server added to it.
2022-10-18T07:05:29.269677Z 0 [ERROR] [MY-010119] [Server] Aborting
2022-10-18T07:05:29.269807Z 0 [System] [MY-010910] [Server] /usr/sbin/mysqld: Shutdown complete (mysqld 8.0.28-0ubuntu0.20.04.3)  (Ubuntu).
``` 

### To Reproduce

The problem can be reproduced sometimes which depends on load on machine and juju.

1. juju add-model openstack
2. juju deploy <https://github.com/hemanthnakkina/snap-sunbeam/blob/main/bundles/etc/bundles/control-plane.yaml> --trust 

### Environment

$ snap list | egrep 'juju|microk8s'
juju                            3.0-rc1                     21020  3.0/beta            canonical**   -
microk8s                        v1.25.2                     4049   1.25-strict/stable  canonical**   -

### Relevant log output

```shell
unit-mysql-0: 12:35:00 DEBUG unit.mysql/0.juju-log Initialising instance
unit-mysql-0: 12:35:13 DEBUG jujuc running hook tool "juju-log" for mysql/0-start-350622731184055142
unit-mysql-0: 12:35:13 DEBUG unit.mysql/0.juju-log Waiting for instance to be ready
unit-mysql-0: 12:35:13 DEBUG jujuc running hook tool "juju-log" for mysql/0-start-350622731184055142
unit-mysql-0: 12:35:13 INFO unit.mysql/0.juju-log Configuring instance
unit-mysql-0: 12:35:13 DEBUG jujuc running hook tool "juju-log" for mysql/0-start-350622731184055142
unit-mysql-0: 12:35:13 DEBUG unit.mysql/0.juju-log Configuring users
unit-mysql-0: 12:35:13 DEBUG jujuc running hook tool "juju-log" for mysql/0-start-350622731184055142
unit-mysql-0: 12:35:13 DEBUG unit.mysql/0.juju-log Configuring instance for InnoDB on mysql-0.mysql-endpoints.openstack.svc.cluster.local
unit-mysql-0: 12:35:16 INFO juju.worker.caasunitterminationworker terminating due to SIGTERM
unit-mysql-0: 12:35:16 DEBUG juju.worker.dependency "caas-unit-termination-worker" manifold worker completed successfully
unit-mysql-0: 12:35:16 ERROR juju.worker.uniter.operation hook "start" (via hook dispatching script: dispatch) failed: signal: terminated
....
unit-mysql-0: 12:35:29 DEBUG jujuc running hook tool "juju-log" for mysql/0-mysql-pebble-ready-5200172316667521259
unit-mysql-0: 12:35:29 DEBUG unit.mysql/0.juju-log Initialising instance
unit-mysql-0: 12:35:29 DEBUG jujuc running hook tool "juju-log" for mysql/0-mysql-pebble-ready-5200172316667521259
unit-mysql-0: 12:35:29 ERROR unit.mysql/0.juju-log Exited with code 1. Stderr:
unit-mysql-0: 12:35:29 DEBUG jujuc running hook tool "juju-log" for mysql/0-mysql-pebble-ready-5200172316667521259
unit-mysql-0: 12:35:29 ERROR unit.mysql/0.juju-log   2022-10-18T07:05:29.267960Z 0 [System] [MY-013169] [Server] /usr/sbin/mysqld (mysqld 8.0.28-0ubuntu0.20.04.3) initializing of server in progress as process 165
unit-mysql-0: 12:35:29 DEBUG jujuc running hook tool "juju-log" for mysql/0-mysql-pebble-ready-5200172316667521259
unit-mysql-0: 12:35:29 ERROR unit.mysql/0.juju-log   2022-10-18T07:05:29.269607Z 0 [ERROR] [MY-010457] [Server] --initialize specified but the data directory has files in it. Aborting.
unit-mysql-0: 12:35:29 DEBUG jujuc running hook tool "juju-log" for mysql/0-mysql-pebble-ready-5200172316667521259
unit-mysql-0: 12:35:29 ERROR unit.mysql/0.juju-log   2022-10-18T07:05:29.269615Z 0 [ERROR] [MY-013236] [Server] The designated data directory /var/lib/mysql/ is unusable. You can remove all files that the server added to it.
unit-mysql-0: 12:35:29 DEBUG jujuc running hook tool "juju-log" for mysql/0-mysql-pebble-ready-5200172316667521259
unit-mysql-0: 12:35:29 ERROR unit.mysql/0.juju-log   2022-10-18T07:05:29.269677Z 0 [ERROR] [MY-010119] [Server] Aborting
unit-mysql-0: 12:35:29 DEBUG jujuc running hook tool "juju-log" for mysql/0-mysql-pebble-ready-5200172316667521259
unit-mysql-0: 12:35:29 ERROR unit.mysql/0.juju-log   2022-10-18T07:05:29.269807Z 0 [System] [MY-010910] [Server] /usr/sbin/mysqld: Shutdown complete (mysqld 8.0.28-0ubuntu0.20.04.3)  (Ubuntu).
unit-mysql-0: 12:35:29 DEBUG jujuc running hook tool "status-set" for mysql/0-mysql-pebble-ready-5200172316667521259
unit-mysql-0: 12:35:29 DEBUG jujuc running hook tool "juju-log" for mysql/0-mysql-pebble-ready-5200172316667521259
unit-mysql-0: 12:35:29 DEBUG unit.mysql/0.juju-log Unable to configure instance: 2022-10-18T07:05:29.267960Z 0 [System] [MY-013169] [Server] /usr/sbin/mysqld (mysqld 8.0.28-0ubuntu0.20.04.3) initializing of server in progress as process 165
2022-10-18T07:05:29.269607Z 0 [ERROR] [MY-010457] [Server] --initialize specified but the data directory has files in it. Aborting.
2022-10-18T07:05:29.269615Z 0 [ERROR] [MY-013236] [Server] The designated data directory /var/lib/mysql/ is unusable. You can remove all files that the server added to it.
2022-10-18T07:05:29.269677Z 0 [ERROR] [MY-010119] [Server] Aborting
2022-10-18T07:05:29.269807Z 0 [System] [MY-010910] [Server] /usr/sbin/mysqld: Shutdown complete (mysqld 8.0.28-0ubuntu0.20.04.3)  (Ubuntu)
```
```


### Additional context

_No response_

**Comment by hemanthnakkina** [comment]
The above scenario of restart of charm containers is triggered by this bug https://bugs.launchpad.net/juju/+bug/1993309.
However the charm needs to handle the above mentioned cases if restart happens for some other reasons during bootstrapping.

Or an action/operation that can unblock the user from the situation.

**Comment by taurus-forever** [comment]
Dear @hemanthnakkina , I would like to reproduce the issue and fix it if it is still reproducible.
Can you please re-share the [test YAML file](https://github.com/hemanthnakkina/snap-sunbeam/blob/main/bundles/etc/bundles/control-plane.yaml) as it is gone (404).

Also the mentioned Juju [bug](https://bugs.launchpad.net/juju/+bug/1993309) is fixed for juju 2.9.37+ and operator has improved logic to handle instabilities in Juju storage.
Is it still reproducible/topical? Thank you!

**Comment by taurus-forever** [comment]
Dear @hemanthnakkina , I am resolving the issue due to lack of reply and inability to reproduce this locally.
Most probably it is fixed by https://github.com/canonical/mysql-k8s-operator/pull/123 and https://github.com/canonical/mysql-k8s-operator/pull/128. Feel free to reopen the issue if it is still reproducible. Tnx!
